src/rpcmining.cpp

```diff
@@ -159,10 +161,39 @@ Value generate(const Array& params, bool fHelp)
             LOCK(cs_main);
             IncrementExtraNonce(pblock, chainActive.Tip(), nExtraNonce);
         }
+
+        // Hash state
+        crypto_generichash_blake2b_state eh_state;
+        eh.InitialiseState(eh_state);
+
+        // I = the block header minus nonce and solution.
+        CEquihashInput I{*pblock};
+        CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+        ss << I;
+
+        // H(I||...
+        crypto_generichash_blake2b_update(&eh_state, (unsigned char*)&ss[0], ss.size());
+
         while (!CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
-            // Yes, there is a chance every nonce could fail to satisfy the -regtest
-            // target -- 1 in 2^(2^256). That ain't gonna happen.
             pblock->nNonce = ArithToUint256(UintToArith256(pblock->nNonce) + 1);
+
+            // H(I||V||...
+            crypto_generichash_blake2b_state curr_state;
+            curr_state = eh_state;
+            crypto_generichash_blake2b_update(&curr_state,
+                                              pblock->nNonce.begin(),
+                                              pblock->nNonce.size());
+
+            // (x_1, x_2, ...) = A(I, V, n, k)
+            std::set<std::vector<unsigned int>> solns = eh.BasicSolve(curr_state);
+
+            for (auto soln : solns) {
+                pblock->nSolution = soln;
+
+                if (CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
```

Comment:
Shouldn't need to, because the solver should only output valid solutions (and if it doesn't, that should be captured by a test).